### ISSUE TITLE
sudo python setup.py install on Debian cannot find data files

### ISSUE THREAD

**Issue by benjaoming** [learningequality/ka-lite#3609]
```
kalite start
Traceback (most recent call last):
  File "/usr/share/kalite/kalitectl.py", line 500, in <module>
    args=arguments['DJANGO_OPTIONS']
  File "/usr/share/kalite/kalitectl.py", line 353, in start
    '--pid-file={0000:s}'.format(PID_FILE_JOB_SCHEDULER)
  File "/usr/share/kalite/kalitectl.py", line 288, in manage
    get_commands()
  File "/usr/share/kalite/python-packages/django/core/management/__init__.py", line 109, in get_commands
    apps = settings.INSTALLED_APPS
  File "/usr/share/kalite/python-packages/django/conf/__init__.py", line 53, in __getattr__
    self._setup(name)
  File "/usr/share/kalite/python-packages/django/conf/__init__.py", line 48, in _setup
    self._wrapped = Settings(settings_module)
  File "/usr/share/kalite/python-packages/django/conf/__init__.py", line 132, in __init__
    mod = importlib.import_module(self.SETTINGS_MODULE)
  File "/usr/share/kalite/python-packages/django/utils/importlib.py", line 35, in import_module
    __import__(name)
  File "/usr/local/lib/python2.7/dist-packages/ka_lite-0.14.dev-py2.7.egg/kalite/settings/__init__.py", line 18, in <module>
    from .base import *
  File "/usr/local/lib/python2.7/dist-packages/ka_lite-0.14.dev-py2.7.egg/kalite/settings/base.py", line 188, in <module>
    shutil.copytree(os.path.join(_data_path, 'locale'), p)
  File "/usr/lib/python2.7/shutil.py", line 171, in copytree
    names = os.listdir(src)
OSError: [Errno 2] No such file or directory: '/usr/share/kalite/locale'
ed@FLE-Debian:~$
```

**Comment by EdDixon** [comment]
to replicate install using the new python setup command "sudo python setup.py install" then cd to your home folder you will find "kalite" command works but "kalite start" will generate the above issue.

**Comment by EdDixon** [comment]
kalite restart, kalite test, kalite shell all fail with the same error

**Comment by benjaoming** [comment]
@ed thanks for the details :) it's okay I think the above traceback would be plentyful for catching this fella.

**Comment by benjaoming** [comment]
well, actually doing a "sudo python setup.py install" works fine on my system, I even removed my "~/.kalite" dir before running it.

I'm gonna ask you to try again once a certain PR is merged...

**Comment by benjaoming** [comment]
@EdDixon would there be a point about assuming that this only happens on Debian and not Ubuntu?

**Comment by EdDixon** [comment]
@benjaoming It is possible that it is not happening on ubuntu if the python-appconf package is installed which is not available to debian or other linux repos or windows python as far as I can tell. It is called as only appconf in what ever script is calling it which I have not been able to find (perhaps a requirement of another package in use?) . I recently submitted a PR for addition of django-appconf to requirements which was ubuntus older means of dealing with django configuration and the the kalite.bat start command which was not finding appconf prior to that additional package installation seems to be working and not returning the "module not found appconf" error it was previously. I do not feel this resolves the real problem however as other similar modules seem to be missing such as compressor instead of django-compressor etc. I feel strongly we need to find the package with these requirements and align them to a more common package across python deployments if possible to maintain the maximum amount of platform compatibility we can?

**Comment by benjaoming** [comment]
@EdDixon what's this appconf? I don't understand what you're talking about here :) When the data files that get installed using setuptools are not found after installing, this is an issue with setuptools or possibly an error that I need to fix..

We try to offload the whole platform circus about where to install stuff and maintain updates through setuptools. Ideally, this would be a beautiful walk in the park!!

Unfortunately setuptools has an inconsistent way of dealing with data files outside of python packages (a python package is a folder that contains an `__init__.py` file and usually that's what you distribute). So when we have stuff like the `content/` and `python-packages/` folders sitting outside of the `kalite` package, they are no longer "package data" but just "data" and installing them stops being trivial. As I'm trying not to mess too much with the current layout, I'm trying to leave things where they are and write a `setup.py` script that just works on top of everything.

Also, what might be confusing is the fact that "setuptools" does not refer to a single library!!! It could be a setuptools (from python-setuptools package) or a pip implementation (pip monkey patches setuptools).

If you're interested, read more here:

http://python-notes.curiousefficiency.org/en/latest/pep_ideas/core_packaging_api.html

**Comment by benjaoming** [comment]
kalite doesn't use an "appconf" library anywhere in the kalite codebase nor do we use it in anything that we shipped in "python-packages"..

**Comment by EdDixon** [comment]
Hi Ben, in trying to use the command you referenced in the meeting "python setup.py install" issues were encountered please check out the testing room for details. As a workaround to allow us to test the windows functionality prior to the installer actually being built we attempted to use kalite.bat as we did the .13.x series however that command also failed referencing the appconf package in windows in the error. When I added the django-appconf package to the requirements, installed the package through pip and reran the command bin\windows\kalite.bat start it was successful in starting the application. A small step forward..  Please join us in the rooms to help us resolve these issues together if possible or to better inform us of changes to the base that would keep these procedures from working as we expect. I looked hard but also could not find any reference in our code to this requirement, so as you say we are not requiring it but yet it is required by something because removing the package caused the exact same command to immediately fail?

**Comment by benjaoming** [comment]
@EdDixon I'm not in the FLE chair this week. I just wanted to keep the ball rolling :) But as I said, there are no references to any "appconf" in our current codebase.. maybe it has something to do with django-compressor that you mentioned was installed separately.. not sure.

Btw. running "python setup.py install" does not automatically install dependencies -- but running "pip install ." (from inside the ka-lite source code dir) does because then it's pip's version of setuptools that's running the setup.py script, and pip will automatically fetch dependencies.

**Comment by EdDixon** [comment]
Thanks, I did discover that it is not installing the requires.txt in testing here but with both modules manually loaded the "python setup.py install" gets as far as "...will configure database and prepare it for use. Press [Enter] to continue"  before immediately dropping back to the shell ? Not giving us any idea why it failed but I am suspecting another missing module. I looked everywhere in our code to references to both appconf and compressor but can not find anywhere where we are actually calling them so I can only assume they are dependencies of an a yet unknown package or perhaps as you suggested another module themselves?

**Comment by benjaoming** [comment]
@EdDixon found out now! :) django-appconf is a requirement of django-compressor and will be automatically installed when you do "pip install -e ."

**Comment by benjaoming** [comment]
> "python setup.py install" gets as far as "...will configure database and prepare it for use. Press [Enter] to continue" before immediately dropping back to the shell ?

Are you sure this happens running setup.py? Don't you mean "kalite manage setup" ?

**Comment by EdDixon** [comment]
Excellent! Sorry but kalite is still not resolving on any platforms as a system command and bin/windows/kalite manage setup fails.. Checking now for the exact error...

**Comment by MCGallaspy** [comment]
> bin/windows/kalite manage setup fails
What error do you get? Is it possible that you meant to type `bin/windows/kalite.bat manage setup`.

**Comment by benjaoming** [comment]
> kalite is still not resolving on any platforms as a system command

huh? it certainly does on my platform :)

**Comment by EdDixon** [comment]
No @MCGallaspy I am referring to the results I am having here on the windows system right now I can try these in debian as well but suspect it will be the same which is a good thing as it means we are coming to a common code base with our installers here which has been the whole point right? @benjaoming it is not on windows or MAC atm when using either the "python setup.py install" or testing pre-builds of the official installers. I will now fire up debian and report back what is happening there but I was able to install it so suspect I can also resolve the kalite command in debian which in linux I would expect to be the easiest of the three to have it resolve via symlink but that was apparently tried and failed on the recent MAC installer?

**Comment by EdDixon** [comment]
On debian as reported here the " python setup.py install" and" sudo python setup.py install" fails however the "bin\kalite manage setup" will setup the system properly resulting ihn the kalite command being available as a system command and everything appears to be functioning as designed. But this does not work on those other platforms? @benjaoming I must be confused in your responce to my question how can we get the kalite command to be available as a system command in these other platforms I thought you told me "python setup.py install" would do that but that is not working even in linux as an installer? What exactly is the purpose of this command then if not meant to be a universal installer for linux/mac/and windows? It does seem to be trying to install on these systems but is failing?

**Comment by benjaoming** [comment]
It can be reliably recreated on Travis, so good news: I can recreate it and I have a test case now :)

https://travis-ci.org/benjaoming/ka-lite/jobs/62079291#L3070

**Comment by benjaoming** [comment]
This was all due to a misunderstanding, I thought that a local directory "locales" I had was part of the distributed code. It's not.

Fixed in #3630

**Comment by aronasorman** [comment]
Is this fixed in your PR @benjaoming?

**Comment by benjaoming** [comment]
It should be, but I would like to hear it from @EdDixon too!

**Comment by EdDixon** [comment]
@ben sorry looks like this one slipped by me just caught it in this email
thread showing a very incomplete conversation but it sounds as though the
issue was fixed! Very glad to hear this and anxious to check the builds
coming with this fix! Thanks,

Ed Dixon
Corrections Adviser
Foundation for Learning Equality <http://learningequality.org/>
Give to FLE, Support FLE via Amazon Smile!
<https://learningequality.org/donate/>
M (575)-636-9549

**Comment by setuppy** [comment]
I'm new biggner on kali linux please solve my erorr for setup.py when i enter these command in linux in terminal  in next i see erorr is (no module named mechanize) wtf what is mechanize

**Comment by MCGallaspy** [comment]
Hi @setuppy, I'm not sure what you're referring to -- is it related to the original issue reported here? If not, please open a new issue, and be sure to indicate what version of KA Lite you're using as well as describe steps to replicate the bug.

**Comment by setuppy** [comment]
Its only happen when i upgrade my python version and i'm deleted my old stored linux and have installed new fresh linux now i'm fine without upgrade any fuck python becouse of these bro ......... but still i'm not fine facebook hacking is tottaly prank witg kali linux .......... what is next way for hacking fb ....... ????
